internal/syncx/workerpool.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package syncx
+
+import (
+	"runtime"
+	"sync"
+)
+
+// WorkerPool is a concurrent task processor, scheduling and running tasks from
+// a source channel, feeding any errors into a sink.
+type WorkerPool[T any, R any] struct {
+	tasks   chan T         // Input channel waiting to consume tasks
+	results chan R         // Result channel for consuers to wait on
+	working sync.WaitGroup // Waitgroup blocking on worker liveness
+}
+
+// NewWorkerPool creates a worker pool with the given number of max task capacity
+// and an optional goroutine count to execute on. If 0 threads are requested, the
+// pool will default to the number of (logical) CPUs.
+func NewWorkerPool[T any, R any](tasks int, threads int, f func(T) R) *WorkerPool[T, R] {
```

Comment:
Hmm, but actually, since your `R` is just `error`, the `WorkerPool` then becomes equivalent to https://pkg.go.dev/golang.org/x/sync/errgroup with a limit, so you could just as well use that:

```go
var pool errgroup.Group
pool.SetLimit(threads)
pool.Go(func () error {
    // commit state object
})
...
pool.Wait()
```

Reply:
`errgroup.Group` might actually be better, since it also cancels all work on the first error.

Reply:
That catch is that that will promote just dumping everything in there, mixing tasks that have different resource bounds. I can. imagine that simple clogging the whole thing because one resource holds off another type.

E.g If I merge CPU and disk tasks, then one thread could block on disk, not doing any CPU processing any more because of it. With more typed pools, it's harder to mix and match (and uglier API to make more than one thing concurrent). Hence my thought with the pipeline...

Basically what we want is CPU to be running concurrently on all cores, but not more to not trash. On the other hand, we want to run disk as parallel as possible, since we're going to be more limited thread wise than the SSD's capacity anyway. So that's a different concurrency than the pool.

Reply:
errgroup is different, that blocks submitting a task when all the threads are busy. also errgroup is a new goroutine for each task vs. this one reusing goroutines to make the creation overhead smaller for very tiny tasks

Reply:
Benchmarked a lot. It was very inconclusive whethre workerpool or errgroup is worth it. So I went with errgroup, since it's something we don't have to maintain. for now at least